File: app/src/main/java/com/example/v_mail/mail/MailService.java
package com.example.v_mail.mail;

import android.os.Build;

import com.example.v_mail.mail.java.Pop3ClientJava;
import com.example.v_mail.mail.java.SmtpClientJava;
import com.example.v_mail.mail.jni.Pop3ClientJNI;
import com.example.v_mail.mail.jni.SmtpClientJNI;

import java.io.IOException;
import java.util.Base64;
import java.util.List;

public class MailService {
    private final SmtpClient smtpClient;
    private final Pop3Client pop3Client;

    public MailService(String smtpServer, int smtpPort, String pop3Server, int pop3Port) {
        this.smtpClient = new SmtpClientJNI(smtpServer, smtpPort);
        this.pop3Client = new Pop3ClientJNI(pop3Server, pop3Port);
    }

    private String encodeBase64(String input) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            return Base64.getEncoder().encodeToString(input.getBytes());
        }
        throw new RuntimeException("SDK Vesion is not correct");
    }

    public void sendEmail(String from, String to, String subject, String body, String username, String password) throws IOException {

        try {
            smtpClient.connect();
            smtpClient.authenticate(encodeBase64(username), encodeBase64(password));
            smtpClient.sendEmail(new Email(from, to, subject, body));
            smtpClient.close();
        }
        finally {
            smtpClient.close();
        }
    }

    public List<Email> getInbox(String username, String password) throws IOException {
        try {
            pop3Client.connect();
            pop3Client.authenticate(username, password);
            List<Email> emails = pop3Client.getInbox();
            return emails;
        }
        catch (Exception e) {
            e.printStackTrace();
            return List.of();
        }
        finally {
            pop3Client.close();
        }
    }

}

File: app/src/main/java/com/example/v_mail/mail/MailService.kt
package com.example.v_mail.mail

import android.content.Context
import android.util.Base64

class MailService private constructor(context: Context) {

    private val sharedPreferences = context.getSharedPreferences("vmail_prefs", Context.MODE_PRIVATE)

    private var pop3Client: Pop3Client? = null
    private var smtpClient: SmtpClient? = null
    private var isSmtpConnected: Boolean = false

    init {
        val usernameBase64 = sharedPreferences.getString("username", null)
        val passwordBase64 = sharedPreferences.getString("password", null)

        if (usernameBase64 != null && passwordBase64 != null) {
            // Декодируем имя пользователя и пароль из Base64
            val username = String(Base64.decode(usernameBase64, Base64.DEFAULT))
            val password = String(Base64.decode(passwordBase64, Base64.DEFAULT))

            pop3Client = Pop3Client()
            smtpClient = SmtpClient()

            // Подключаем POP3 и SMTP с использованием декодированных данных
            pop3Client?.connect("pop3.example.com", 110)
            pop3Client?.authenticate(username, password)

            smtpClient?.connect()
            smtpClient?.authenticate(username, password)
        }
    }

    // Подключение к POP3 серверу
    fun connectPop3(port: Int) {
        pop3Client?.connect("pop3.example.com", port)
    }

    // Получение списка сообщений
    fun listMessages(): Array<String>? {
        return pop3Client?.listEmails()
    }

    // Подключение к SMTP серверу
    fun connectSmtp(): Boolean {
        try {smtpClient?.connect(); }
        catch (e: Exception) {
            isSmtpConnected = false
            return false
        }
        isSmtpConnected = true
        return true
    }

    // Отправка письма
    fun sendEmail(from: String, to: String, subject: String, body: String): Boolean {
        try {
            if (!isSmtpConnected) {
                connectSmtp();
            }

            smtpClient?.sendEmail(from, to, subject, body)
        }
        catch (e: Exception) {
            return false;
        }
        return true
    }

    // Отключение от SMTP сервера
    fun disconnectSmtp() {
        smtpClient?.close()
        isSmtpConnected = false
    }

    // Отключение от POP3 сервера
    fun disconnectPop3() {
        pop3Client?.quit()
    }

    companion object {
        private var instance: MailService? = null

        // Метод для получения единственного экземпляра MailService
        fun getInstance(context: Context): MailService {
            if (instance == null) {
                instance = MailService(context)
            }
            return instance!!
        }
    }
}
